### ISSUE TITLE
Running into a wried discount issue with data lose using GroundDB 

### ISSUE THREAD

**Issue by almogdesign** [GroundMeteor/db#94]
So were running into some wried discount issue with GroundDB, if our app goes from offline to online and their is a internet disconnection were seeing the app refresh and the local data that was saved via GroundDB is lost and was never upload. 

It seems that the app clears the local storage before it was uploaded. 

Has anyone else had this issue?

**Comment by rafaelquintanilha** [comment]
Hard to tell without seeing any code. You mean your data is always lost when you have a disconnection or there is any other case when it works properly?

**Comment by almogdesign** [comment]
Yes I get a internet disconnection the app goes into offline mode everything is fine but then when the app connects again it refreshes deletes the current local data and re saves. It's overwriting the local data. 

The code is based on the examples provided but I am using iron-router

**Comment by joaopiopedreira** [comment]
Hi [almogdesign](https://github.com/almogdesign), in order not to loose previously saved data in the client, you might want to try to ground your collections in this way (according to the documentation):

    var groundList = new Ground.Collection('list', {
       cleanupLocalData: false
    });

This way, your client data will not be overwritten and only changed records will be added/removed/changed.

If you have large collections on the client, you might want to try [this fix](https://github.com/GroundMeteor/db/pull/95). It worked for me.

Bear in mind that after the app refreshes (offline to online transition), you loose all you Session variables and are thus not able to bring the user to where he was prior to the reload. To overcome that, I started using persistent Session vars (take a look [here](https://atmospherejs.com/u2622/persistent-session)). At least now I can take the user exactly where he was before the reload, which it good (had a lot of complaints before this change...). I use the "auth" version for persistence: Session vars are kept since a the user logs in and removed once he logs out. Very helpful! 

Also, watch out for the Ground.ready() hook. When online, it works fine but if you are offline it hangs and it never reports "true".

I'm currently using [GroundDB](https://github.com/GroundMeteor/db) together with mdg's [appcache](https://atmospherejs.com/meteor/appcache) in a web/mobile app. It's been a challenge to control all this page reload / data re-fetch behaviour but I think I'm getting there. I'm happy to exchange my experience with you in case you need further assistance.

My ultimate challenge now is to solve this very strange issue (with cordova):
- I build my app with "meteor build" and make an apk file;
- Install this file in an android phone and it all works fine;
- I deploy a new version of the app (with only minor changes, like a space in a file) to my server and try to connect with the cordova app;
- The app reports an offline status (from Meteor.status()) and never recovers from that.

Now, how strange is that?

Cheers and good luck!

**Comment by raix** [comment]
*create a package for appcache only adding it when on the `web` platform - don't use it on cordova..*

**Comment by almogdesign** [comment]
@joaopiopedreira this is really great thanks for the help will be testing it out today. Bye the way you can use amplifyjs for the sessions.

**Comment by almogdesign** [comment]
@raix @joaopiopedreira I'm not sure the issue is due to cordova, we also this issue with certain tablets. The tablet would go offline and once there was a wifi connection it would not go back online. It was Samsung Tab tablets were still looking into this.

**Comment by joaopiopedreira** [comment]
@raix, but how does a Cordova app behaves when offline without appcache? Will it still work? 

Are there any improvements in the works in order to make appcache play nicer with Cordova?

Thanks

Enviado do meu iPhone

No dia 04/05/2015, às 06:41, Morten N.O. Nørgaard Henriksen <notifications@github.com> escreveu:

**Comment by almogdesign** [comment]
@joaopiopedreira can we connect on Skype / email would like to learn more about your experience. I think we're both doing the same and running into the same issues

**Comment by joaopiopedreira** [comment]
Sure Almog, if today is a good day for you, it's fine for me (bank holiday in the UK). I'm doing some catch-up work in the hotel.

Skype name: joaopiopedreira

Cheers,

Enviado do meu iPhone

No dia 04/05/2015, às 10:29, Almog Koren <notifications@github.com> escreveu:

**Comment by almogdesign** [comment]
@rafaelquintanilha great

**Comment by joaopiopedreira** [comment]
Hi @raix, would you tell me how to do that (create a package only for a certain platform)?

Many thanks,

Enviado do meu iPhone

No dia 04/05/2015, às 06:41, Morten N.O. Nørgaard Henriksen <notifications@github.com> escreveu:

**Comment by raix** [comment]
```js
Package.describe({
  name: 'browserappcache',
  summary: 'Only add appcache to browsers'
});

Package.onUse(function(api) {
  api.use('appcache', ['web.browser', 'server']);
});

```

**Comment by joaopiopedreira** [comment]
Thank you @raix, appreciated :)

One more, pretty please:

I have two large collections that do not change much (server side) and are never modified client side.

Do you think it's a good idea to copy them into a client only collection on start up and ground them as such (to avoid ground:db having to subscribe to them on every reconnect, erasing and recreating them?

Thanks

Enviado do meu iPhone

No dia 06/05/2015, às 20:11, Morten N.O. Nørgaard Henriksen <notifications@github.com> escreveu:

**Comment by raix** [comment]
@joaopiopedreira if the two collections never change you could have something like:
```js
  var foo = new Ground.Collection('foo', { cleanupLocalData: false });

  Meteor.startup(function() {
    if (!foo.findOne()) {
      Meteor.subscribe('foo', function() {
        // Manually triggering a clean up of local only data
        groundList.removeLocalOnly(); 
      });
    } else {
      // Got data already
    }
  });
```

**Comment by joaopiopedreira** [comment]
Got it! Thank you very much @raix :-)
